compiler/src/dotty/tools/dotc/transform/FirstTransform.scala

```diff
@@ -144,7 +144,14 @@ class FirstTransform extends MiniPhase with InfoTransformer { thisPhase =>
   }
 
   override def transformIdent(tree: Ident)(implicit ctx: Context): Tree =
-    if (tree.isType) toTypeTree(tree) else constToLiteral(tree)
+    if (tree.isType) {
+      toTypeTree(tree)
+    } else if (tree.name != nme.WILDCARD) {
+      // We want to constant-fold _some_ idents here - for instance, @switch needs to see literals in patterns.
+      // However, constant-foldable wildcards can occur in patterns, for instance as `case _: "a"`;
```

Comment:
I don't understand why we need to handle `case _: "a"` specially but not `case x: "a"`

Reply:
Oh, I guess because the latter is desugared to `case x @ (_: "a")` ? This is all pretty subtle :/

Reply:
Argh, I see where the comment is hilariously unclear. Give me a second.